File: code/csharp/ch06/luavm/binchunk/BinaryChunk.cs
﻿namespace luavm.binchunk
{
    struct binaryChunk
    {
        header header;
        byte sizeUpvalues;
        Prototype mainFunc;
    }

    struct header
    {
        byte[] signature;
        byte version;
        byte format;
        byte[] luacData;
        byte cintSize;
        byte sizetSize;
        byte instructionSize;
        byte luaIntegerSize;
        byte luaNumberSize;
        long luacInt;
        double luacNum;
    }

    public struct Prototype
    {
        internal string Source;
        internal uint LineDefined;
        internal uint LastLineDefined;
        internal byte NumParams;
        internal byte IsVararg;
        internal byte MaxStackSize;
        internal uint[] Code;
        internal object[] Constants;
        internal Upvalue[] Upvalues;
        internal Prototype[] Protos;
        internal uint[] LineInfo;
        internal LocVar[] LocVars;
        internal string[] UpvalueNames;
    }

    internal struct Upvalue
    {
        internal byte Instack;
        internal byte Idx;
    }

    internal struct LocVar
    {
        internal string VarName;
        internal uint StartPC;
        internal uint EndPC;
    }


    internal static class BinaryChunk
    {
        public const string LUA_SIGNATURE = "\x1bLua";
        public const byte LUAC_VERSION = 0x53;
        public const byte LUAC_FORMAT = 0;
        public const string LUAC_DATA = "\x19\x93\r\n\x1a\n";
        public const uint CINT_SIZE = 4;
        public const uint CSIZET_SIZE_32 = 4;
        public const uint CSIZET_SIZE_64 = 8;
        public const uint INSTRUCTION_SIZE = 4;
        public const uint LUA_INTEGER_SIZE = 8;
        public const uint LUA_NUMBER_SIZE = 8;
        public const ushort LUAC_INT = 0x5678;
        public const double LUAC_NUM = 370.5;

        public const byte TAG_NIL = 0x00;
        public const byte TAG_BOOLEAN = 0x01;
        public const byte TAG_NUMBER = 0x03;
        public const byte TAG_INTEGER = 0x13;
        public const byte TAG_SHORT_STR = 0x04;
        public const byte TAG_LONG_STR = 0x14;


        public static Prototype Undump(byte[] data)
        {
            var reader = new Reader {data = data};
            reader.checkHeader();
            reader.readByte();
            return reader.readProto("");
        }
    }
}

File: code/python/src/bin_chunk/binary_chunk.py
from bin_chunk.binary_reader import BinaryReader
from bin_chunk.binary_chunk_header import BinaryChunkHeader
from vm.prototype import Prototype


class BinaryChunk:
    def __init__(self, chunk):
        self.binary_reader = BinaryReader(chunk)
        self.header = None
        self.size_upvalues = None
        self.main_func = None

    @staticmethod
    def is_binary_chunk(data):
        if data is None or len(data) < 4:
            return False

        for i in range(4):
            if data[i] != BinaryChunkHeader.LUA_SIGNATURE[i]:
                return False

        return True

    def print_header(self):
        self.header.dump()

    def check_header(self):
        self.header.check()

    def print_main_func(self):
        self.main_func.dump()

    def get_main_func(self):
        return self.main_func

    def undump(self):
        self.header = BinaryChunkHeader(self.binary_reader)
        self.check_header()
        self.size_upvalues = self.binary_reader.read_uint8()
        self.main_func = Prototype()
        self.main_func.init_from_br(self.binary_reader, '')
        return self.main_func
